### PR TITLE
Adding to_string methods

### PR DESCRIPTION
**Pull request by not-Ian**
Adding to_string methods
Also added some build items to .gitignore 

Re: #3451

### CODE REVIEW

File: nano/node/common.cpp [hunk 1]
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
 	}
 }
 
+std::string nano::keepalive::to_string ()
+{
+	std::stringstream stream;
+
+	stream << "\n" + header.to_string () + "\n";
+
+	for (auto peer (peers.begin ()), j (peers.end ()); peer != j; ++peer)
```
Comment (dsiganos) [inline nano/node/common.cpp]: Use:
```
 auto peer = peers.begin ()
 ```

File: nano/node/common.cpp [hunk 2]
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
 	}
 }
 
+std::string nano::keepalive::to_string ()
+{
+	std::stringstream stream;
+
+	stream << "\n" + header.to_string () + "\n";
```
Comment (dsiganos) [inline nano/node/common.cpp]: Why is there a leading new line? That looks weird. It should be up to the caller to put leading or trailing whitespace. This code should use minimal whitespace.

File: nano/node/common.cpp [hunk 3]
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
 	}
 }
 
+std::string nano::keepalive::to_string ()
```
Comment (dsiganos) [inline nano/node/common.cpp]: This function can be const.

File: .gitignore [hunk 4]
```diff
@@ -57,6 +57,9 @@ Release
 *.aps
 *.db
 *.opendb
+/out/**
```
Comment (dsiganos) [inline .gitignore]: Let's not commit changes to .gitingore that are not relevant to this PR.

File: nano/node/common.cpp [hunk 5]
```diff
@@ -996,6 +1011,17 @@ bool nano::frontier_req::operator== (nano::frontier_req const & other_a) const
 	return start == other_a.start && age == other_a.age && count == other_a.count;
 }
 
+std::string nano::frontier_req::to_string ()
```
Comment (dsiganos) [inline nano/node/common.cpp]: remove leading space, this could be compressed to one line without whitespace if you shorted the message.

E.g.
```
start=xxxx age=yyyy count=zzzz
```

Try not to avoid whitespace, ideally use bio vertical whitespace.
Use vertical whitespace only where it adds a lot of value.

File: nano/node/common.cpp [hunk 6]
```diff
@@ -1189,6 +1215,15 @@ void nano::telemetry_req::visit (nano::message_visitor & visitor_a) const
 	visitor_a.telemetry_req (*this);
 }
 
+std::string nano::telemetry_req::to_string ()
```
Comment (dsiganos) [inline nano/node/common.cpp]: This could be a one-liner
```
return header.to_string()
```

File: nano/node/common.cpp [hunk 7]
```diff
@@ -1221,6 +1256,15 @@ void nano::telemetry_ack::serialize (nano::stream & stream_a) const
 	}
 }
 
+std::string nano::telemetry_ack::to_string () const
+{
+	std::stringstream stream;
+
```
Comment (dsiganos) [inline nano/node/common.cpp]: could be a one-liner

File: nano/node/common.hpp [hunk 8]
```diff
@@ -424,6 +426,7 @@ class telemetry_req final : public message
 	void serialize (nano::stream &) const override;
 	bool deserialize (nano::stream &);
 	void visit (nano::message_visitor &) const override;
+	std::string to_string ();
```
Comment (dsiganos) [inline nano/node/common.hpp]: this can be marked as const

File: nano/node/common.hpp [hunk 9]
```diff
@@ -368,6 +369,7 @@ class frontier_req final : public message
 	uint32_t age;
 	uint32_t count;
 	static std::size_t constexpr size = sizeof (start) + sizeof (age) + sizeof (count);
+	std::string to_string ();
```
Comment (dsiganos) [inline nano/node/common.hpp]: this can be marked as const

File: nano/node/common.hpp [hunk 10]
```diff
@@ -311,6 +311,7 @@ class keepalive final : public message
 	bool operator== (nano::keepalive const &) const;
 	std::array<nano::endpoint, 8> peers;
 	static std::size_t constexpr size = 8 * (16 + 2);
+	std::string to_string ();
```
Comment (dsiganos) [inline nano/node/common.hpp]: this can be marked as const

File: nano/node/common.cpp (hunk 6 above)
Comment (dsiganos) [inline nano/node/common.cpp]: this can be marked as const

**Review by dsiganos — CHANGES_REQUESTED** [review]

File: nano/node/common.cpp (hunk 7 above)
Comment (not-Ian) [reply nano/node/common.cpp]: are you thinking just drop the stringstream object entirely? I'm only using it because I saw it was used in a separate to_string function. Is it only useful if you're formatting multiple variables together or something?

File: nano/node/common.cpp (hunk 3 above)
Comment (not-Ian) [reply nano/node/common.cpp]: marked const in most recent commit

File: nano/node/common.hpp (hunk 8 above)
Comment (not-Ian) [reply nano/node/common.hpp]: marked const in most recent commit

File: nano/node/common.hpp (hunk 9 above)
Comment (not-Ian) [reply nano/node/common.hpp]: marked const in most recent commit

File: nano/node/common.hpp (hunk 10 above)
Comment (not-Ian) [reply nano/node/common.hpp]: marked const in most recent commit

File: nano/node/common.cpp (hunk 6 above)
Comment (not-Ian) [reply nano/node/common.cpp]: marked const in most recent commit

File: nano/node/common.cpp (hunk 7 above)
Comment (not-Ian) [reply nano/node/common.cpp]: I also made telemetry_req::to_string() a one liner. It was very similar

Comment (dsiganos) [reply nano/node/common.cpp]: string stream is a flexible stream buffer that can accumulate lots of data in a flexible way and you can output formatting rules to format your data. It is convenient.
But in this case, your data is ready, no processing needed, just pass it along as it is.

File: nano/node/common.cpp [hunk 11]
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
 	}
 }
 
+std::string nano::keepalive::to_string () const
+{
+	std::stringstream stream;
+
+	stream << header.to_string () + "\n";
+
+	for (auto peer = peers.begin (), j (peers.end ()); peer != j; ++peer)
```
Comment (dsiganos) [inline nano/node/common.cpp]: Please use
```
j = peers.end()
```

we prefer the assignment style of initialisation

Comment (not-Ian) [reply nano/node/common.cpp]: if I see somewhere else that's not doing the assignment style of initialization, can I change it, or would that be out of the scope of this PR and not approved of

**Comment by not-Ian** [comment]
Example outputs:

nano::confirm_req:

`NetID: 5243(live), VerMaxUsingMin: 18/18/18, MsgType: 4(confirm_req), Extensions: 1100`

---

nano::telemetry_ack:

```
{
    "block_count": "0",
    "cemented_count": "0",
    "unchecked_count": "0",
    "account_count": "0",
    "bandwidth_cap": "0",
    "peer_count": "0",
    "protocol_version": "0",
    "uptime": "0",
    "genesis_block": "0000000000000000000000000000000000000000000000000000000000000000",
    "major_version": "0",
    "minor_version": "0",
    "patch_version": "0",
    "pre_release_version": "0",
    "maker": "0",
    "timestamp": "0",
    "active_difficulty": "0000000000000000"
}
```

---

nano::frontier_req:

`Start account: 0000000000000000000000000000000000000000000000000000000000000000 | Maximum age of account: 1647681972 | Maximum number of accounts to include: 1048576`

---

nano::confirm_ack:

`Account: A30E0A32ED41C8607AA9212843392E853FCBCB4E7CB194E35C94F07F91DE59EF`

---

nano::telemetry_req:

`NetID: 5243(live), VerMaxUsingMin: 19/19/18, MsgType: 12(telemetry_req), Extensions: 0000`

**Comment by not-Ian** [comment]
And the rest of them:

node::node_id_handshake:

```
NetID: 5243(live), VerMaxUsingMin: 18/18/18, MsgType: 10(node_id_handshake), Extensions: 0003
Query cookie: 794DE15DC2B3CB84CE0AF3917F5B3A29C5219F83BD0B8BDB84A293C064716C18
Account: 1A2C0156415D1386FB86D03CD99966E73C1DBBC0985D9D86C484F7C554D00308 | 349660F5768BA07E479C76CE5F4E44189325A84F9A2DE6CF23E4AF9D40D7A98344589EAAC3A738D1240901B77C2AE35A063A8AA1E056DA7E22D1BD365DD0CC0D
```

---

node::publish

```
{
    "type": "state",
    "account": "nano_1chan9rmpx8wxh83wcmsxf4ywcaqq5riimq8d3876iouzdjk9nsn4cf56qn4",
    "previous": "67950A8DD4B95F3E201885DF54FCF3CDDFBA572F8332E3EDE6D4CB39A3E89DF1",
    "representative": "nano_3bsnis6ha3m9cepuaywskn9jykdggxcu8mxsp76yc3oinrt3n7gi77xiggtm",
    "balance": "99858859438706100489898610573647",
    "link": "8D4B2671F35B03656D97881E75878D8E0BF374ACC61A92FD04735F6DFE4E097A",
    "link_as_account": "nano_35cd6srz8pr5eopsh41ygp5ru5idyftcsjitkdyiawtzfqz6w4dth7338pr7",
    "signature": "0A605F86F14D0EA9A475F614DAEC4E270FE9FB38229F7332F86184B00008DCAADEB97BFFE0001D68AA3C4F3466B0F6305CD67D6DBB01F5D30DD9F25ABB6BF403",
    "work": "24578869cb8c5048"
}
```

---

node::bulk_pull:

`Start account: 03489141285CDF3936090F1F6E68238FEA79F419E5C4F15C3E969A7E26B28DC2 | End block hash: 74CD5BBA03C45481424183F2E68FA812FEAD481D5D5B509B4B576FEBDBA52B1C | Count: 0`

---

node::bulk_pull_account:

`Account: FD98692F080CED721C8449E034F3D8ADC11C890CE46B208D8F3CC73A765D54EB | Minimum amount: 000000000000D3C21BCECCEDA1000000 | Pending hash and amount`

---

node::bulk_push:

`NetID: 5243(live), VerMaxUsingMin: 19/19/18, MsgType: 7(bulk_push), Extensions: 0000NetID: 5243(live), VerMaxUsingMin: 19/19/18, MsgType: 7(bulk_push), Extensions: 0000Socket error: connect, 10061`

File: nano/node/common.cpp (hunk 11 above)
Comment (dsiganos) [reply nano/node/common.cpp]: You can make small obvious changes but bigger changes should be done in a separate PR

**Comment by dsiganos** [comment]
It is good to keep logs of this nature as concise as possible.
I would change this:
```
Start account: 0000000000000000000000000000000000000000000000000000000000000000 | Maximum age of account: 1647681972 | Maximum number of accounts to include: 1048576
```

to this:
```
0000000000000000000000000000000000000000000000000000000000000000 maxage=1647681972 count=1048576
```

**Comment by not-Ian** [comment]
gonna work on tests next. And also clean up the two functions that are returning JSON

File: .gitignore (hunk 4 above)
Comment (not-Ian) [reply .gitignore]: I have deleted these changes

File: nano/node/common.cpp (hunk 1 above)
Comment (not-Ian) [reply nano/node/common.cpp]: done. Should I make this change in other places in the same file where they use `auto peer (peers.begin ())` as opposed to `auto peer = peers.begin ()`? Like, in `nano::keepalive::serialize` around line 690

File: nano/node/common.cpp (hunk 2 above)
Comment (not-Ian) [reply nano/node/common.cpp]: done

File: nano/node/common.cpp (hunk 5 above)
Comment (not-Ian) [reply nano/node/common.cpp]: pushed just now

File: nano/node/common.cpp (hunk 1 above)
Comment (dsiganos) [reply nano/node/common.cpp]: In general yes, but for now since this is your first PR, let's just focus on getting it through the review process and merged.

File: nano/node/messages.cpp [hunk 12]
```diff
@@ -707,6 +707,21 @@ bool nano::keepalive::deserialize (nano::stream & stream_a)
 	return error;
 }
 
+std::string nano::keepalive::to_string () const
+{
+	std::stringstream stream;
+
+	stream << header.to_string ();
+
+	for (auto peer = peers.begin (), peers_end = peers.end (); peer != peers_end; ++peer)
```
Comment (dsiganos) [inline nano/node/messages.cpp:716]: There is no need to take a copy of `peers_end = peers.end ();` to avoid calling the function repeatedly. compilers know how to optimise these things easily. The code should be written for readability first, not for optimisation.

This is better because it is easily readable and because the compiler will make the optimisation anyway:
```
for (auto peer = peers.begin (); peer != peers.end(); ++peer)
```

File: nano/slow_test/vote_cache.cpp [hunk 13]
```diff
@@ -215,7 +215,7 @@ TEST (vote_cache, perf_multithreaded)
 	// Ensure our generated votes go to inactive vote cache instead of active elections
 	node.active.clear ();
 
-	run_parallel (thread_count, [&node, &reps, &blocks] (int index) {
+	run_parallel (thread_count, [&node, &reps, &blocks, &vote_count, &single_vote_size, &single_vote_reps] (int index) {
```
Comment (dsiganos) [inline nano/slow_test/vote_cache.cpp:218]: This seems unrelated.

File: nano/node/messages.cpp [hunk 14]
```diff
@@ -1581,4 +1680,20 @@ std::size_t nano::node_id_handshake::size (nano::message_header const & header_a
...
 	}
 	return result;
+}
+
+std::string nano::node_id_handshake::to_string () const
+{
+	std::stringstream stream;
+
+	stream << header.to_string () + "\n";
+	stream << "cookie=" + query->to_string ();
```
Comment (dsiganos) [inline nano/node/messages.cpp:1690]: The query is optional, so you need to check it has a value before trying to access it.

File: nano/slow_test/vote_cache.cpp (hunk 13 above)
Comment (not-Ian) [reply nano/slow_test/vote_cache.cpp:218]: I originally needed this change to get tests to run

Comment (dsiganos) [reply nano/slow_test/vote_cache.cpp:218]: OK

File: nano/node/messages.cpp [hunk 15]
```diff
@@ -1287,6 +1381,11 @@ bool nano::telemetry_ack::is_empty_payload () const
 	return size () == 0;
 }
 
+std::string nano::telemetry_ack::to_string () const
+{
+	return data.to_string ();
```
Comment (dsiganos) [inline nano/node/messages.cpp:1386]: It is possible for telemetry data to have empty payload. You need to check is_empty_payload.

File: nano/core_test/message.cpp [hunk 16]
```diff
@@ -276,3 +276,242 @@ TEST (message, bulk_pull_serialization)
 	ASSERT_FALSE (error);
 	ASSERT_TRUE (header.bulk_pull_ascending ());
 }
+
+TEST (message, keepalive_to_string)
+{
+	nano::work_thresholds work_threshold = nano::work_thresholds (0, 0, 0);
```
Comment (dsiganos) [inline nano/core_test/message.cpp:282]: You can create the keepalive object in one line like this:
```
nano::keepalive keepalive = nano::keepalive (nano::dev::network_params.network);
```

**Comment by not-Ian** [comment]
Closing in favor of smaller, individual PRs
